src/aggregator/aggregator/generic_elem.go

```diff
@@ -123,84 +130,59 @@ type GenericElem struct {
-	sp policy.StoragePolicy,
-	aggTypes maggregation.Types,
-	pipeline applied.Pipeline,
-	numForwardedTimes int,
-	idPrefixSuffixType IDPrefixSuffixType,
-	opts Options,
-) (*GenericElem, error) {
+// NewGenericElem returns a new GenericElem.
+func NewGenericElem(data ElemData, opts Options) (*GenericElem, error) {
 	e := &GenericElem{
 		elemBase: newElemBase(opts),
 		values:   make([]timedAggregation, 0, defaultNumAggregations), // in most cases values will have two entries
 	}
-	if err := e.ResetSetData(id, sp, aggTypes, pipeline, numForwardedTimes, idPrefixSuffixType); err != nil {
+	if err := e.ResetSetData(data); err != nil {
 		return nil, err
 	}
 	return e, nil
 }
 
-// MustNewGenericElem creates a new element, or panics if the input is invalid.
-func MustNewGenericElem(
-	id id.RawID,
-	sp policy.StoragePolicy,
-	aggTypes maggregation.Types,
-	pipeline applied.Pipeline,
-	numForwardedTimes int,
-	idPrefixSuffixType IDPrefixSuffixType,
-	opts Options,
-) *GenericElem {
-	elem, err := NewGenericElem(id, sp, aggTypes, pipeline, numForwardedTimes, idPrefixSuffixType, opts)
+// MustNewGenericElem returns a new GenericElem and panics if an error occurs.
+func MustNewGenericElem(data ElemData, opts Options) *GenericElem {
+	elem, err := NewGenericElem(data, opts)
 	if err != nil {
 		panic(fmt.Errorf("unable to create element: %v", err))
 	}
 	return elem
 }
 
 // ResetSetData resets the element and sets data.
-func (e *GenericElem) ResetSetData(
-	id id.RawID,
-	sp policy.StoragePolicy,
-	aggTypes maggregation.Types,
-	pipeline applied.Pipeline,
-	numForwardedTimes int,
-	idPrefixSuffixType IDPrefixSuffixType,
-) error {
-	useDefaultAggregation := aggTypes.IsDefault()
+func (e *GenericElem) ResetSetData(data ElemData) error {
+	useDefaultAggregation := data.AggTypes.IsDefault()
 	if useDefaultAggregation {
-		aggTypes = e.DefaultAggregationTypes(e.aggTypesOpts)
+		data.AggTypes = e.DefaultAggregationTypes(e.aggTypesOpts)
 	}
-	if err := e.elemBase.resetSetData(id, sp, aggTypes, useDefaultAggregation, pipeline, numForwardedTimes, idPrefixSuffixType); err != nil {
+	if err := e.elemBase.resetSetData(data, useDefaultAggregation); err != nil {
 		return err
 	}
-	if err := e.typeSpecificElemBase.ResetSetData(e.aggTypesOpts, aggTypes, useDefaultAggregation); err != nil {
+	if err := e.typeSpecificElemBase.ResetSetData(e.aggTypesOpts, data.AggTypes, useDefaultAggregation); err != nil {
 		return err
 	}
 	// If the pipeline contains derivative transformations, we need to store past
 	// values in order to compute the derivatives.
 	if !e.parsedPipeline.HasDerivativeTransform {
 		return nil
 	}
-	numAggTypes := len(e.aggTypes)
-	if cap(e.lastConsumedValues) < numAggTypes {
-		e.lastConsumedValues = make([]transformation.Datapoint, numAggTypes)
-	}
-	e.lastConsumedValues = e.lastConsumedValues[:numAggTypes]
-	for i := 0; i < len(e.lastConsumedValues); i++ {
-		e.lastConsumedValues[i] = transformation.Datapoint{Value: nan}
-	}
 	return nil
```

Comment:
Looks like since there's no initialization work here, there's no need to check `e.parsedPipeline.HasDerivativeTransform`?

i.e. should just be:
```go
return e.typeSpecificElemBase.ResetSetData(e.aggTypesOpts, aggTypes, useDefaultAggregation)
```